### PR TITLE
Context menu enhancements

### PR DESCRIPTION
**Pull request by mathieumg**
Thanks @justinnichols for providing the context menu with more structure and in cleaner fashion than what I had hacked together in https://github.com/mathieumg/plaidchat/commit/b47e6285f109a13a2c3e1c0b75b33ef809413966

However, there are a couple things I thought were worth salvaging:

* Hyperlinks can be many things, they should not be limited to HTTP/HTTPS and absolute URLs. I removed the `http`/`https` check and replaced it with a check that prepends the current team's URL if the link is relative to the domain. (i.e. starts with `/`) Since `slack-context-menu` is used entirely statically, I had to pass along the team URL when calling its `handleRightClick` method. 
* To be more in line with the official Slack client, I added the ability to copy an hyperlink and renamed the menu options to "Copy Link" and "Open Link". I can provide a screenshot from the Windows client if you want.

This now allows, for example, to copy/open "ftp://" URLs and more importantly copy/open permalinks to messages, which is something I personally use a lot to quote previous messages. Both of these work that way in the official client.

I didn't squash the changes yet as I thought you might not want all of them, so let me know.

### CODE REVIEW

**Comment by justinnichols** [comment]
I think this is a great enhancement, thanks for doing so!  I appreciate you feeling this is a clean method to build upon.  I hope to continue helping to enhance it as well.

Assuming the linting check has passed (I'm currently unable to give it a go, but can try later this evening) and @twolfson agrees with the change, all is well from my end for you to squash the commits for it to be pulled in.

File: app/components/slack-context-menu.js [hunk 1]
```diff
@@ -108,17 +108,28 @@
...
-			if (hrefValue.startsWith('http') || hrefValue.startsWith('https')) {
-				menu.append(new gui.MenuItem({
-					label: 'Open Link in Browser',
-					click: function () {
-						gui.Shell.openExternal(hrefValue);
-					}
-				}));
+
+			// Relative URL, let's prepend the team URL.
+			if (hrefValue.startsWith('/')) {
```
Comment (twolfson) [inline app/components/slack-context-menu.js]: Sometimes URLs can be `//slack.com/hi` which uses the same protocol as the existing page. I tried using `url.parse` to detect `hostname` but it looks like that won't work:

http://requirebin.com/?gist=86f3a3a987c9779dec0b

However, I wrote a module a while back for work that seems to line up with this use case. Maybe we should use that:

https://github.com/underdogio/resolve-link

File: app/components/slack-context-menu.js [hunk 2]
```diff
@@ -138,7 +149,7 @@
 	}
 
 	module.exports = {
-		handleRightClick: function (window, target, x, y) {
+		handleRightClick: function (window, target, x, y, teamUrl) {
```
Comment (twolfson) [inline app/components/slack-context-menu.js]: These parameters are getting a bit unwieldy. Maybe we should convert them into an object for easier passing?

```js
// params = {window, target, x, y, teamUrl}
function (params) {
```

Comment (twolfson) [inline app/components/slack-context-menu.js]: Then in the later invocations, we can always have the syntax `build*Menu(menu, params);`

**Comment by twolfson** [comment]
The gist of the changes looks good. Left some notes with respect to URL edge cases and more maintainable code

File: app/components/slack-context-menu.js (hunk 1 above)
Comment (twolfson) [inline app/components/slack-context-menu.js]: Upon re-evaluating the test cases for `resolve-link`, it looks like we never tested for `//` either. Maybe it's best to change this `if` to use `hrefValue.startsWith('/') && !hrefValue.startsWith('//')`

Comment (mathieumg) [inline app/components/slack-context-menu.js]: `resolve-link` seems nice, do you want me to leverage it in this PR?

File: app/components/slack-context-menu.js (hunk 2 above)
Comment (mathieumg) [inline app/components/slack-context-menu.js]: I felt the same (as pointed out in my initial message) but didn't want to change the approach without prior discussion. Eventually, the `team` prop is something that could be interesting to have as React context instead!

I can change the arguments to use a parameters object instead.

File: app/components/slack-context-menu.js (hunk 1 above)
Comment (twolfson) [inline app/components/slack-context-menu.js]: Nah, it feels like we would be forcing a hexagon into a circular hole. Probably best to use the `startsWith` versions for now.

File: app/components/slack-context-menu.js (hunk 2 above)
Comment (twolfson) [inline app/components/slack-context-menu.js]: I don't quite follow what you mean; `teamUrl` is already coming from a React context =/

Comment (mathieumg) [inline app/components/slack-context-menu.js]: https://github.com/plaidchat/plaidchat/blob/master/app/components/slack-window.js#L243 looks like a standard prop! :P I was talking about https://medium.com/@learnreact/context-f932a9abab0e but they're not documented yet, hence the eventually.

Comment (twolfson) [inline app/components/slack-context-menu.js]: Ah, k. Neat

File: app/components/slack-context-menu.js (hunk 1 above)
Comment (mathieumg) [inline app/components/slack-context-menu.js]: Hadn't seen your last message before mine, I'll update the condition with what you suggested for the time being.

**Comment by mathieumg** [comment]
Let me know if that's what you had in mind. Did you want me to convert the other handlers as well?

File: app/components/slack-context-menu.js [hunk 3]
```diff
@@ -108,17 +108,31 @@
...
 			var hrefValue = target.attributes.href.value;
-			if (hrefValue.startsWith('http') || hrefValue.startsWith('https')) {
-				menu.append(new gui.MenuItem({
-					label: 'Open Link in Browser',
-					click: function () {
-						gui.Shell.openExternal(hrefValue);
-					}
-				}));
+
+			// Relative URL, let's prepend the team URL.
```
Comment (twolfson) [inline app/components/slack-context-menu.js]: Can we provide an example comment of what the new value will be?

```js
// e.g. `/customize` -> `https://my.slack.com/customize`
```

**Comment by twolfson** [comment]
Yep, that's what I had in mind. Let's definitely move it to the other methods -- it will allow us to get rid of the gnarly destructuring here:

https://github.com/plaidchat/plaidchat/pull/139/files#diff-3fd39051cf5ff554092fd129e27f9e3eR156

**Comment by twolfson** [comment]
Also, please squash the commits for this PR =/

**Comment by mathieumg** [comment]
I was wondering if I was allowed to use ES2015 destructuring there, but I guess https://github.com/plaidchat/plaidchat/blob/master/package.json#L36 means that I cannot. :P I'll simply redirect the params directly and manually destructure `target` then?

As for the squashing, I certainly intended to do so before the merge, but didn't see the hurry. (See OP)

**Comment by twolfson** [comment]
Yep, stick to ES5 for now please =)

My complaint was mostly that we broke down `params` and then re-defined the same exact object before passing it along:

https://github.com/plaidchat/plaidchat/pull/139/files#diff-3fd39051cf5ff554092fd129e27f9e3eR156

https://github.com/plaidchat/plaidchat/pull/139/files#diff-3fd39051cf5ff554092fd129e27f9e3eR175

With respect to squashing, I wanted to make sure you didn't forget ;)

**Comment by twolfson** [comment]
Side note/pro tip: Either link to a `git` sha or tag when mentioning code in PR's. Linking to the `master` branch is inevitably volatile and ruins the link when re-reviewing conversation history =/

That is, replace the:

https://github.com/plaidchat/plaidchat/blob/master/package.json#L36

with either:

https://github.com/plaidchat/plaidchat/blob/b484bdfa588ce90a28c64d5b21df33965fc22785/package.json#L36

^ retrieved via pressing `y`

https://github.com/plaidchat/plaidchat/blob/v2.14.1/package.json#L36

^ retrieved by using the Branch selection at the top and using Tags

**Comment by mathieumg** [comment]
> ^ retrieved via pressing y

I had been looking for this all my life. Thank you. Seriously. I even had in mind the idea to make a Chrome extension for this... :\ I'm aware of the volatility problem but since fetching the blob hash manually and replacing it was cumbersome I didn't always do it or always thought about it. I honestly think GitHub should use the hash by default.

**Comment by twolfson** [comment]
Glad to be of help =) Fun fact: Even more GitHub shortcuts can be found via `?` (contents change depending on page)

Also, here's where I found out about it originally: http://zachholman.com/talk/git-github-secrets/

**Comment by mathieumg** [comment]
Changes done & squashed!

File: app/components/slack-window.js [hunk 4]
```diff
@@ -44,7 +44,7 @@
 			var that = this;
 
 			win.addEventListener('contextmenu', function handleContextMenu (evt) {
-				SlackContextMenu.handleRightClick(that.getWindow(), evt.target, evt.x, evt.y);
+				SlackContextMenu.handleRightClick({window: that.getWindow(), target: evt.target, x: evt.x, y: evt.y, teamUrl: that.props.team.team_url});
```
Comment (twolfson) [inline app/components/slack-window.js]: This line is a bit lengthy. Maybe we can create a `var params` with a couple of line feeds on the previous line to shorten it?

```js
var params = {
    target: evt.target, x: evt.x, y: evt.y,
    window: that.getWindow(), teamUrl: that.props.team.team_url
};
SlackContextMenu.handleRightClick(params);
```

**Comment by twolfson** [comment]
Looks good but the commit message isn't explaining the commit too well =/ It feels like everything is part of the body. Maybe we can update it to something like:

```
Added link copying and relative link support
```

This should be the first line in a squash commit message. However, it can also be updated via `git commit -m "Message" --amend`.

File: app/components/slack-context-menu.js (hunk 3 above)
Comment (twolfson) [inline app/components/slack-context-menu.js]: This comment has been resolved :+1:

File: app/components/slack-window.js (hunk 4 above)
Comment (mathieumg) [inline app/components/slack-window.js]: What about simply:

```js
SlackContextMenu.handleRightClick({
    target: evt.target, 
    teamUrl: that.props.team.team_url,
    x: evt.x, 
    y: evt.y,
    window: that.getWindow()
});
```

**Comment by mathieumg** [comment]
Just to clarify, do you want me to prepend the commit message with that line or to replace it with that line?

Thanks.

File: app/components/slack-window.js (hunk 4 above)
Comment (twolfson) [inline app/components/slack-window.js]: Sure, works for me

**Comment by twolfson** [comment]
I will leave that choice up to you. I mostly care about commit messages, I don't usually dig into bodies when perusing `git log` or the GitHub commit log

**Comment by mathieumg** [comment]
How does that look?

**Comment by twolfson** [comment]
Commit looks good. It seems like we didn't handle the last comment though =/

**Comment by mathieumg** [comment]
My bad, juggling from projects/computers, missed that. Should be good now.

**Comment by twolfson** [comment]
Yep, looks good. Going to wait for Travis CI to approve and then land it.

**Comment by twolfson** [comment]
This has been landed/released in `2.15.0`. I am going to add you as a collaborator as well :+1:

**Comment by justinnichols** [comment]
Thanks for the great enhancements!

**Comment by mathieumg** [comment]
Thanks for the review! :+1:
